// @bairong/jax/src/index.js
// import 'babel-polyfill';
import Layout from './components/layout';
import Home from './views/home/home';
import Error404 from './views/errorPage/404';
import RuleAddEdit from './components/rules/RuleAddEdit';
import PresetRuleChange from './components/rules/PresetRuleChange';
import TableMergeCell from './components/table/TableMergeCell';
import Authorization from './components/authorization';

import utils from './utils';
// import utils from './libs/utils/regExp';
// import http from './libs/http';

import user from './store/modules/user';
import app from './store/modules/app';

const components = {
  Layout,
  Home,
  Error404,
  TableMergeCell,
  RuleAddEdit,
  PresetRuleChange,
  Authorization
};

const jax = {
  user,
  app,
  ...utils,
  ...components
};
// const jax = Object.assign({}, { user }, utils, components);

// 定义 install 方法，接收 Vue 作为参数。如果使用 use 注册插件，则所有的组件都将被注册
const install = function (Vue) {
  // 判断是否安装
  if (install.installed) return;
  // 遍历注册全局组件
  components.map(component => Vue.component(component.name, component));
};

// auto install
if (typeof window !== 'undefined' && window.Vue) {
  install(window.Vue);
}

// export {Layout, Home, Error404, PBreadcrumb};
// module.exports.default = module.exports = jax;
export { Layout, Home, Error404, user, app };
export default jax;
